core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsTransport.java

```diff
@@ -1547,6 +1546,9 @@ public Object up(Message msg) {
       public void up(MessageBatch batch) {
          ByRef<Message> firstMessage = new ByRef<>(null);
          batch.forEach((message, messages) -> {
+            // Removed messages are null
+            if (message == null) return;
+
             // Regular (non-OOB) messages must be processed in-order
             // Normally a batch should either have only OOB or only regular messages,
             // but we check for every message to be on the safe side.
```

Comment:
What about using `org.infinispan.util.concurrent.BlockingManager#limitedBlockingExecutor()` with `concurrency=1` for the regular messages?

Reply:
I'm not sure that would help... first off, the delivery should happen on a non-blocking thread, because everything but the sending of the reply should be non-blocking.

Second, UNICAST3/NAKACK2 use a Table as a queue for regular messages, which are only delivered after all previous messages have been delivered. If we pretend the regular message has been delivered and we put it in a limited executor's queue instead, it allows JGroups to remove the message faster, but I don't think it reduces the amount of work done by UNICAST3/NAKACK2.

PS: I already forgot I said I'd wait for the perf test results and I already pushed a change to process all messages in the delivery thread, not just regular messages :)